spring-integration-core/src/main/java/org/springframework/integration/handler/DelayHandler.java

```diff
@@ -365,9 +367,16 @@ private void doReleaseMessage(Message<?> message) {
 
 	private boolean removeDelayedMessageFromMessageStore(Message<?> message) {
 		if (this.messageStore instanceof SimpleMessageStore) {
-			SimpleMessageGroup messageGroup =
-					(SimpleMessageGroup) this.messageStore.getMessageGroup(this.messageGroupId);
-			return messageGroup.remove(message);
+			synchronized (this.messageGroupId) {
+				Collection<Message<?>> messages = this.messageStore.getMessageGroup(this.messageGroupId).getMessages();
+				if (messages.contains(message)) {
+					this.messageStore.removeMessageFromGroup(this.messageGroupId, message);
```

Comment:
We should use the new `removeMessagesFromGroup()` method - this one returns the whole group and then we discard it - ok for a simple store but not for a persistent store.

In fact, this method should be deprecated - this is the only place in the framework (aside from tests) that this method is called.